### PR TITLE
Fix iscsi volume mount issue

### PR DESCRIPTION
**Pull request by surajssd**
- Create iscsi initiator name on workers

  - Enable `iscsid` service on baremetal.
  
  - With new Flatcar `2605.9.0` the `/etc/iscsi` directory is shipped and the initiator name file (`/etc/iscsi/initiatorname.iscsi`) has incomplete config which causes the `iscsid` process to start with warning that the initiator name is improper. This PR makes sure that the initiator name is proper. More explanation: https://github.com/kinvolk/lokomotive/issues/1257#issuecomment-743050474.
  
  - This PR adds a `ExecPreStart` which creates `iscsid` initiator name file.
  
  - Remove deprecated `enable: true` for the existing `iscsid` service configuration.
  
  - We can use the script `iscsi-gen-initiatorname` but it creates the initiator name with FQDN `suse.de` in it, so this script creates the initiator name with FQDN with kinvolk.io in it.

- kubelet image: Add Dockerfile

  This PR adds a Dockerfile and a script which runs `iscsiadm` in host namespace.

- Replace image from quay.io/poseidon/kubelet to quay.io/kinvolk/kubelet

- kubelet: Remove mount of `iscsiadm` binary from host

  This PR removes the mount of `iscsiadm` from the host, expecting the image to ship `iscsiadm`.

- tinkerbell: Update kubelet to v1.19.4

  There was a regression from previous Kubernetes update to `v1.19.4`. This PR fixes that regression.

---

Fixes: #1257 
Fixes: #447

### CODE REVIEW

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 1]
```diff
@@ -100,6 +100,9 @@ systemd:
         ExecStop=/etc/kubernetes/delete-node
         [Install]
         WantedBy=multi-user.target
+    - name: "iscsid.service"
+      enabled: true
+      enable: true
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:120]: why both?

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 2]
```diff
@@ -100,6 +100,9 @@ systemd:
         ExecStop=/etc/kubernetes/delete-node
         [Install]
         WantedBy=multi-user.target
+    - name: "iscsid.service"
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:118]: We don't need quotes?

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 3]
```diff
@@ -100,6 +100,19 @@ systemd:
...
+    - name: "create-iscsi-initiatorname-file.service"
+      enable: true
+      contents: |
+        [Unit]
+        Description=Create ISCSI initiator name file
+        Before=iscsid.service
+        [Service]
+        Type=oneshot
+        RemainAfterExit=true
+        ExecStart=/opt/create-iscsi-initiatorname
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:114]: Will it run on every boot? I guess that may have some side effects? Why can't we specify a drop-in for `iscsid` service to run this before start?

Can we have a command inline here, rather than a script file? For one linters, it's less redirection.

File: assets/charts/control-plane/kubelet/templates/kubelet-ds.yaml [hunk 4]
```diff
@@ -129,6 +129,11 @@ spec:
         - name: iscsiadm
           mountPath: /usr/sbin/iscsiadm
           readOnly: false
+        # iscsi libraries from the host
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/templates/kubelet-ds.yaml:132]:
```suggestion
# iSCSI libraries from the host.
```

File: assets/charts/control-plane/kubelet/templates/kubelet-ds.yaml [hunk 5]
```diff
@@ -194,6 +199,15 @@ spec:
       - name: iscsiadm
         hostPath:
           path: /usr/sbin/iscsiadm
+      # iscsi libraries from the host
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/templates/kubelet-ds.yaml:202]:
```suggestion
# iSCSI libraries from the host.
```

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl [hunk 6]
```diff
@@ -82,6 +82,8 @@ systemd:
           -v /opt/cni/bin:/opt/cni/bin:ro \
           -v /usr/lib/os-release:/etc/os-release:ro \
           -v /usr/sbin/iscsiadm:/usr/sbin/iscsiadm:rw \
+          -v /lib64/libisns.so.0:/usr/lib/x86_64-linux-gnu/libisns.so.0 \
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:85]: Are those libraries present in kubelet image we use? I wonder if this override may affect some other binaries in the image.

**Review by invidian — CHANGES_REQUESTED** [review]

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 6 above)
Comment (pothos) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:85]: And anyway this will break next week (earlier or later) when the dependencies change. Please use one approach I suggested in https://github.com/kinvolk/Flatcar/issues/293#issuecomment-742671025 (I think the wrapper script with `nsenter` is quite feasible).

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 1 above)
Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:120]: Copy paste from Packet. Fixing it.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 2 above)
Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:118]: Copy paste from Packet. Fixing it.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 3 above)
Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:114]: We can't modify/create "service" config when the service is already on the verge of starting.

Like if I create a dropin for `iscsid` and add the oneline script in `ExecStartPre` then it does not load the new config. We need someone else doing it.

Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:114]: Why it does not load the new config? Actually, it might also be what is happening with kubelet at the moment, which might be causing #1191 :thinking:

**Review by invidian — CHANGES_REQUESTED** [review]
I'd give a try to @pothos idea with a wrapper script instead of mounting host libraries, that seems like a cleaner solution.

**Comment by pothos** [comment]
There is no guarantee that a file like `libopeniscsiusr.so.0.2.0` exists on the host, unless you turn off OS updates.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 6 above)
Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:85]: @pothos the problem with using nsenter trick is that we don't call iscsiadm explicitly, kubelet does that. Would you suggest running kubelet with nsenter?

I think using the `ldd` trick is feasible.

Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:85]: @surajssd I think it should be enough to create a file called `iscsiadm` in `kubelet` container which is a script instead of a host binary. Probably we should write it using Ignition and then bind-mount it.

I wouldn't run entire `kubelet` with `nsenter`.

Comment (pothos) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:85]: Yes, the kubelet can't be run with nsenter because the kubelet file does not exist on the host. It's about a wrapper script called `iscsiadm` which would call it.

Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: @invidian @pothos could you give it a try now? What do you think of the current approach.

The kubelet systemd process's `ExecStartPre` copies the libs to a certain dir which is mounted inside the Kubelet pod and this Kubelet pod copies those files before starting the Kubelet process.

Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: @surajssd what is wrong with the wrapper script approach? Current approach seems like a hammer...

Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: It worked, created a new image. I guess I was trying to fix using something that didn't need creating a new image. But anyway, I have created an image with the new script.

Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: if we would have it here, it would be better documented. Do we have a repository for image building?

Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Do you mean I should add something more apart from the README?

Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Ah, I didn't see the README. But why can't we deploy a script using Ignition?

BTW, for kubelet images, it would be nice if we could build a multi-arch manifests instead of having separate tags.

Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: How would ignition put the file inside kubelet pod. Yes we can do multiple volume mounts. But doesn't this provide us more control over image now and smaller ignition which has been a reason for failure on AWS?

Added multi arch image building instructions.

File: assets/charts/control-plane/kubelet/container-image/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,3 @@
+FROM quay.io/poseidon/kubelet:v1.19.4
+
+COPY iscsiadm /bin/iscsiadm
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/container-image/Dockerfile:3]:
```suggestion
COPY ./iscsiadm /bin/
```

File: assets/charts/control-plane/kubelet/container-image/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
+# Lokomotive Kubelet Image
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/container-image/README.md:1]: Do we have some docs about updating Kubernetes version? It would be nice if we could like to this file from there. I think `assets/charts/control-plane/kubelet/container-image/README.md` will be difficult to find. Also, this image is not only used by the chart but also in Ignition files, so honestly I'm not sure if this is the right place for it...

File: assets/charts/control-plane/kubelet/container-image/README.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+This image is built on top of the Typhoon Kubelet image. This image includes a wrapper script that runs `isciadm` in the host namespace. This ensures that all the `isciadm` libraries are loaded from the host.
+
+## Updating Image Tag
+
+In this directory run the following command to update the Kubernetes version:
+
+```bash
+export CURRENT_VERSION=
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/container-image/README.md:10]: Current version could be taken from the file itself.

File: assets/charts/control-plane/kubelet/container-image/README.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+```bash
+ARCH=amd64 docker build -t quay.io/kinvolk/kubelet:v1.19.4-amd64 --build-arg ARCH . \
+    && docker push quay.io/kinvolk/kubelet:v1.19.4-amd64
+```
+
+### ARM
+
+```bash
+ARCH=arm64 docker build -t quay.io/kinvolk/kubelet:v1.19.4-arm64 --build-arg ARCH . \
+    && docker push quay.io/kinvolk/kubelet:v1.19.4-arm64
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/container-image/README.md:31]: This is recommended line-breaking, at least in `Dockerfile`.
```suggestion
ARCH=arm64 docker build -t quay.io/kinvolk/kubelet:v1.19.4-arm64 --build-arg ARCH . && \
docker push quay.io/kinvolk/kubelet:v1.19.4-arm64
```

File: assets/charts/control-plane/kubelet/container-image/README.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+sed -i "s|$CURRENT_VERSION|$NEW_VERSION|g" Dockerfile README.md
+```
+
+## Building Image
+
+### x86
+
+```bash
+ARCH=amd64 docker build -t quay.io/kinvolk/kubelet:v1.19.4-amd64 --build-arg ARCH . \
+    && docker push quay.io/kinvolk/kubelet:v1.19.4-amd64
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/container-image/README.md:24]:
```suggestion
ARCH=amd64 docker build -t quay.io/kinvolk/kubelet:v1.19.4-amd64 --build-arg ARCH . && \
docker push quay.io/kinvolk/kubelet:v1.19.4-amd64
```

File: assets/charts/control-plane/kubelet/container-image/README.md [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+```bash
+ARCH=arm64 docker build -t quay.io/kinvolk/kubelet:v1.19.4-arm64 --build-arg ARCH . \
+    && docker push quay.io/kinvolk/kubelet:v1.19.4-arm64
+```
+
+### Combined image tag
+
+Now make sure you have `"experimental": "enabled"` in your
+`~/.docker/config.json` (surrounded by `{` and `}` if the file is otherwise
+empty).
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/container-image/README.md:38]: I wonder if this is still needed with latest Docker.

File: assets/charts/control-plane/kubelet/container-image/README.md [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+```bash
+sed -i "s|$CURRENT_VERSION|$NEW_VERSION|g" Dockerfile README.md
+```
+
+## Building Image
+
+### x86
+
+```bash
+ARCH=amd64 docker build -t quay.io/kinvolk/kubelet:v1.19.4-amd64 --build-arg ARCH . \
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/container-image/README.md:23]: Can't we use env variables in those commands? Then we don't have to replace anything. Just set one variable at the beginning. It would make everything much simpler.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 3 above)
Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:111]: I'm still interested in resolving that.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 14]
```diff
@@ -138,9 +137,21 @@ systemd:
...
+        Description=Create ISCSI initiator name file
+        Before=iscsid.service
+        [Service]
+        Type=oneshot
+        RemainAfterExit=true
+        ExecStart=/opt/create-iscsi-initiatorname
+        [Install]
+        WantedBy=multi-user.target
+        RequiredBy=iscsid.service
+    - name: iscsid.service
```
Comment (invidian) [inline assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:153]: If we keep the quotes as they were we make diff more readable.

**Review by invidian — CHANGES_REQUESTED** [review]

File: assets/charts/control-plane/kubelet/container-image/README.md (hunk 12 above)
Comment (surajssd) [reply assets/charts/control-plane/kubelet/container-image/README.md:38]: Yep we need it: https://docs.docker.com/engine/reference/commandline/manifest_create/

File: assets/charts/control-plane/kubelet/container-image/README.md (hunk 10 above)
Comment (surajssd) [reply assets/charts/control-plane/kubelet/container-image/README.md:31]: These commands are meant to be copy pasted (not for `Dockerfile`), the `\` is going to be overridden and everything will be on one line.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 14 above)
Comment (surajssd) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:153]: Removed that commit.

**Comment by surajssd** [comment]
@invidian 

> I'm still interested in resolving that.

what do you mean?

File: assets/charts/control-plane/kubelet/container-image/README.md (hunk 8 above)
Comment (surajssd) [reply assets/charts/control-plane/kubelet/container-image/README.md:1]: We don't have a document for updating k8s. The instructions are close to `Dockerfile`.

**Comment by invidian** [comment]
>what do you mean?

I means my previous comment:

> Why it does not load the new config? Actually, it might also be what is happening with kubelet at the moment, which might be causing #1191 thinking

According to systemd documentation `ExecStart` should wait for `ExecStartPre` to finish successfully, so I would like to understand, why we can't use `ExecStartPre`.

File: assets/charts/control-plane/kubelet/container-image/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,3 @@
+FROM quay.io/poseidon/kubelet:v1.19.4
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/container-image/Dockerfile:1]: Can we fork github.com/poseidon/kubelet and put this there then?

File: assets/charts/control-plane/kubelet/container-image/README.md (hunk 8 above)
Comment (invidian) [reply assets/charts/control-plane/kubelet/container-image/README.md:1]: Can we fork github.com/poseidon/kubelet and put this there then?

Comment (surajssd) [reply assets/charts/control-plane/kubelet/container-image/README.md:1]: Isn't this essentially the same?

Comment (invidian) [reply assets/charts/control-plane/kubelet/container-image/README.md:1]: I think it's easier to find by mapping `quay.io/kinvolk/kubelet` -> `github.com/kinvolk/kubelet` than to `github.com/kinvolk/lokomotive/assets/charts/control-plane/kubelet/container-image`.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 16]
```diff
@@ -100,6 +100,13 @@ systemd:
         ExecStop=/etc/kubernetes/delete-node
         [Install]
         WantedBy=multi-user.target
+    - name: iscsid.service
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:102]: Please apply that to all platforms, so we remain consistent.

File: assets/charts/control-plane/kubelet/templates/kubelet-ds.yaml [hunk 17]
```diff
@@ -125,10 +125,6 @@ spec:
         - name: etc-resolv
           mountPath: /etc/resolv.conf
           readOnly: true
-        # This is mounted so that storage works fine.
```
Comment (invidian) [inline assets/charts/control-plane/kubelet/templates/kubelet-ds.yaml:128]: IMO this should be committed together with kubelet image change, otherwise checking out this commit will cause a regression.

Alternatively, please swap the commit order so we first replace the image, then we remove the bind mount. This won't create a regressing revision.

File: assets/charts/control-plane/kubelet/container-image/README.md (hunk 8 above)
Comment (invidian) [reply assets/charts/control-plane/kubelet/container-image/README.md:1]: Also, having forked repo would resolve #221.

**Review by invidian — CHANGES_REQUESTED** [review]
Overall LGTM. My main concerns are lack of consistency across the platforms, which I think we should fix before merging. About placement of `Dockerfile`, I'm curious what other think about it.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 18]
```diff
@@ -5,7 +5,11 @@ systemd:
       enable: true
     - name: iscsid.service
       enable: true
-      enabled: true
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:8]: nit: seems like `enable`/`enabled` duplication could be removed in a separate commit.

File: assets/terraform-modules/azure/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 19]
```diff
@@ -7,7 +7,11 @@ systemd:
       mask: true
     - name: iscsid.service
       enable: true
-      enabled: true
```
Comment (invidian) [inline assets/terraform-modules/azure/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:10]: Where is it documented that `enable` should be used instead of `enabled` ?

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 20]
```diff
@@ -5,7 +5,11 @@ systemd:
       enable: true
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:5]: The commit title and message are no longer accurate. It is applied on all platforms.

**Review by invidian — CHANGES_REQUESTED** [review]

File: assets/terraform-modules/azure/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 19 above)
Comment (surajssd) [reply assets/terraform-modules/azure/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:10]: https://coreos.com/ignition/docs/latest/configuration-v2_4.html

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 18 above)
Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:8]: Added the change info to the commit.

File: assets/terraform-modules/azure/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 19 above)
Comment (invidian) [reply assets/terraform-modules/azure/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:10]: Not sure if this documentation is a good reference, as it's from CoreOS, however it says that `enable` is deprecated and `enabled` should be used. I think we should remove the other one (and while doing so we can commit it separately :wink:)

Comment (surajssd) [reply assets/terraform-modules/azure/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:10]: My bad, good catch. I did exact opposite :facepalm: Fixed it now.

**Review by invidian — APPROVED** [review]
LGTM, I'm sorry for so many change requests @surajssd :pray:

**Comment by surajssd** [comment]
@invidian no problem. Thanks for the feedback @invidian and that script trick @pothos. :pray:
